File: tests/bindings/e2e/cpp/worker_long_running_fixture.cpp
#include "common/memory/backend.hpp"
#include "common/runtime/workflow_store.hpp"
#include "worker/workflow_runner.hpp"
#include "worker/workflows/provision_service/handlers.hpp"
#include "worker/workflows/provision_service/registry.hpp"

#include <atomic>
#include <chrono>
#include <stdexcept>
#include <thread>

class CountingWorkflowStore final : public statespec::backend::runtime::RuntimeWorkflowStore
{
  public:
    statespec::backend::WorkflowExecutionRecord keep_alive_step(
        statespec::backend::IBackend& backend,
        const statespec::backend::KeepAliveWorkflowStepRequest& request
    ) override
    {
        keep_alive_count_.fetch_add(1, std::memory_order_relaxed);
        return RuntimeWorkflowStore::keep_alive_step(backend, request);
    }

    int keep_alive_count() const
    {
        return keep_alive_count_.load(std::memory_order_relaxed);
    }

  private:
    std::atomic<int> keep_alive_count_{0};
};

class LongRunningWorkflowStepHandler final : public statespec_generated::worker::workflows::
                                                 provision_service::ProvisionServiceV1StepHandler
{
  public:
    statespec_generated::worker::WorkflowStepResult handle_validate_request(
        statespec::backend::ITransaction& tx,
        const statespec_generated::worker::WorkflowStepHandlerContext& context
    ) override
    {
        if (!tx.is_open())
        {
            throw std::runtime_error("workflow handler received a closed transaction");
        }
        if (context.workflow_name != "ProvisionService" || context.step_name != "validate_request")
        {
            throw std::runtime_error("unexpected workflow step");
        }
        std::this_thread::sleep_for(std::chrono::milliseconds{850});
        handled_ = true;
        return statespec_generated::worker::WorkflowStepResult::complete();
    }

    statespec_generated::worker::WorkflowStepResult handle_create_remote_service(
        statespec::backend::ITransaction& tx,
        const statespec_generated::worker::WorkflowStepHandlerContext&
    ) override
    {
        if (!tx.is_open())
        {
            throw std::runtime_error("workflow handler received a closed transaction");
        }
        return statespec_generated::worker::WorkflowStepResult::fail(
            "unexpected create_remote_service step"
        );
    }

    statespec_generated::worker::WorkflowStepResult handle_wait_for_remote_service(
        statespec::backend::ITransaction& tx,
        const statespec_generated::worker::WorkflowStepHandlerContext&
    ) override
    {
        if (!tx.is_open())
        {
            throw std::runtime_error("workflow handler received a closed transaction");
        }
        return statespec_generated::worker::WorkflowStepResult::fail(
            "unexpected wait_for_remote_service step"
        );
    }

    bool handled() const
    {
        return handled_;
    }

  private:
    bool handled_ = false;
};

int main()
{
    statespec::backend::memory::InMemoryBackend backend;
    CountingWorkflowStore workflows;

    workflows.register_definition(
        backend, statespec::backend::RegisterWorkflowDefinitionRequest{
                     statespec::backend::WorkflowDefinition{
                         "ProvisionService",
                         1,
                         "validate_request",
                         std::chrono::seconds{60},
                         false,
                         {
                             statespec::backend::WorkflowStepDefinition{
                                 "validate_request", std::chrono::seconds{1}, 3
                             },
                             statespec::backend::WorkflowStepDefinition{
                                 "create_remote_service", std::chrono::seconds{1}, 3
                             },
                             statespec::backend::WorkflowStepDefinition{
                                 "wait_for_remote_service", std::chrono::seconds{1}, 3
                             },
                         },
                         statespec::backend::Json::object({}),
                     },
                 }
    );
    workflows.start(
        backend, statespec::backend::StartWorkflowRequest{
                     "wf-long",
                     "ProvisionService",
                     1,
                     "validate_request",
                     statespec::backend::Json::object({}),
                 }
    );

    LongRunningWorkflowStepHandler handler;
    statespec_generated::worker::WorkflowStepInvokerMap invokers;
    statespec_generated::worker::workflows::provision_service::register_workflow_step_invokers(
        invokers, handler
    );
    statespec_generated::worker::WorkflowRunner runner{
        backend, workflows, invokers, "ProvisionWorker", std::chrono::seconds{1}, 3,
    };
    const auto completed = runner.run_once("wf-long", "ProvisionService", 1);
    if (!handler.handled() || !completed.has_value() || completed->status != "Completed")
    {
        throw std::runtime_error("long-running workflow step did not complete");
    }
    if (workflows.keep_alive_count() < 2)
    {
        throw std::runtime_error("long-running workflow step did not receive periodic keepalive");
    }
}

File: tests/bindings/e2e/rust/worker_long_running_fixture.rs
use std::sync::atomic::{AtomicUsize, Ordering};
use std::thread;
use std::time::Duration;

use statespec_generated::backend::{Backend, BackendResult, Transaction};
use statespec_generated::json::Json;
use statespec_generated::memory_backend::InMemoryBackend;
use statespec_generated::runtime_workflows::RuntimeWorkflowStore;
use statespec_generated::workflow::{
    CancelWorkflowRequest, ClaimWorkflowStepRequest, CompleteWorkflowStepRequest,
    FailWorkflowStepRequest, KeepAliveWorkflowStepRequest, RegisterWorkflowDefinitionRequest,
    StartWorkflowRequest, WorkflowDefinition, WorkflowDefinitionRegistration,
    WorkflowExecutionRecord, WorkflowStepDefinition, WorkflowStore,
};
use statespec_generated::workflow_provision_service_handlers::ProvisionServiceV1StepHandler;
use statespec_generated::workflow_provision_service_registry;
use statespec_generated::workflow_runner::WorkflowRunner;
use statespec_generated::workflow_step_handlers::{
    WorkflowStepHandlerContext, WorkflowStepInvokerMap, WorkflowStepResult,
};

struct CountingWorkflowStore {
    inner: RuntimeWorkflowStore,
    keep_alive_count: AtomicUsize,
}

impl CountingWorkflowStore {
    fn new() -> Self {
        Self {
            inner: RuntimeWorkflowStore::new(),
            keep_alive_count: AtomicUsize::new(0),
        }
    }

    fn keep_alive_count(&self) -> usize {
        self.keep_alive_count.load(Ordering::Relaxed)
    }
}

impl WorkflowStore<InMemoryBackend> for CountingWorkflowStore {
    fn register_definition(
        &self,
        backend: &InMemoryBackend,
        request: &RegisterWorkflowDefinitionRequest,
    ) -> BackendResult<WorkflowDefinitionRegistration> {
        self.inner.register_definition(backend, request)
    }

    fn register_definition_tx(
        &self,
        tx: &mut <InMemoryBackend as Backend>::Tx,
        request: &RegisterWorkflowDefinitionRequest,
    ) -> BackendResult<WorkflowDefinitionRegistration> {
        <RuntimeWorkflowStore as WorkflowStore<InMemoryBackend>>::register_definition_tx(
            &self.inner,
            tx,
            request,
        )
    }

    fn inspect_definition(
        &self,
        backend: &InMemoryBackend,
        workflow_name: &str,
        workflow_version: i64,
    ) -> BackendResult<Option<WorkflowDefinition>> {
        self.inner
            .inspect_definition(backend, workflow_name, workflow_version)
    }

    fn inspect_definition_tx(
        &self,
        tx: &mut <InMemoryBackend as Backend>::Tx,
        workflow_name: &str,
        workflow_version: i64,
    ) -> BackendResult<Option<WorkflowDefinition>> {
        <RuntimeWorkflowStore as WorkflowStore<InMemoryBackend>>::inspect_definition_tx(
            &self.inner,
            tx,
            workflow_name,
            workflow_version,
        )
    }

    fn start(
        &self,
        backend: &InMemoryBackend,
        request: &StartWorkflowRequest,
    ) -> BackendResult<WorkflowExecutionRecord> {
        self.inner.start(backend, request)
    }

    fn start_tx(
        &self,
        tx: &mut <InMemoryBackend as Backend>::Tx,
        request: &StartWorkflowRequest,
    ) -> BackendResult<WorkflowExecutionRecord> {
        <RuntimeWorkflowStore as WorkflowStore<InMemoryBackend>>::start_tx(&self.inner, tx, request)
    }

    fn claim_steps(
        &self,
        backend: &InMemoryBackend,
        request: &ClaimWorkflowStepRequest,
    ) -> BackendResult<Vec<WorkflowExecutionRecord>> {
        self.inner.claim_steps(backend, request)
    }

    fn claim_steps_tx(
        &self,
        tx: &mut <InMemoryBackend as Backend>::Tx,
        request: &ClaimWorkflowStepRequest,
    ) -> BackendResult<Vec<WorkflowExecutionRecord>> {
        <RuntimeWorkflowStore as WorkflowStore<InMemoryBackend>>::claim_steps_tx(
            &self.inner,
            tx,
            request,
        )
    }

    fn keep_alive_step(
        &self,
        backend: &InMemoryBackend,
        request: &KeepAliveWorkflowStepRequest,
    ) -> BackendResult<WorkflowExecutionRecord> {
        self.keep_alive_count.fetch_add(1, Ordering::Relaxed);
        self.inner.keep_alive_step(backend, request)
    }

    fn keep_alive_step_tx(
        &self,
        tx: &mut <InMemoryBackend as Backend>::Tx,
        request: &KeepAliveWorkflowStepRequest,
    ) -> BackendResult<WorkflowExecutionRecord> {
        <RuntimeWorkflowStore as WorkflowStore<InMemoryBackend>>::keep_alive_step_tx(
            &self.inner,
            tx,
            request,
        )
    }

    fn complete_step(
        &self,
        backend: &InMemoryBackend,
        request: &CompleteWorkflowStepRequest,
    ) -> BackendResult<WorkflowExecutionRecord> {
        self.inner.complete_step(backend, request)
    }

    fn complete_step_tx(
        &self,
        tx: &mut <InMemoryBackend as Backend>::Tx,
        request: &CompleteWorkflowStepRequest,
    ) -> BackendResult<WorkflowExecutionRecord> {
        <RuntimeWorkflowStore as WorkflowStore<InMemoryBackend>>::complete_step_tx(
            &self.inner,
            tx,
            request,
        )
    }

    fn fail_step(
        &self,
        backend: &InMemoryBackend,
        request: &FailWorkflowStepRequest,
    ) -> BackendResult<WorkflowExecutionRecord> {
        self.inner.fail_step(backend, request)
    }

    fn fail_step_tx(
        &self,
        tx: &mut <InMemoryBackend as Backend>::Tx,
        request: &FailWorkflowStepRequest,
    ) -> BackendResult<WorkflowExecutionRecord> {
        <RuntimeWorkflowStore as WorkflowStore<InMemoryBackend>>::fail_step_tx(
            &self.inner,
            tx,
            request,
        )
    }

    fn cancel(
        &self,
        backend: &InMemoryBackend,
        request: &CancelWorkflowRequest,
    ) -> BackendResult<WorkflowExecutionRecord> {
        self.inner.cancel(backend, request)
    }

    fn cancel_tx(
        &self,
        tx: &mut <InMemoryBackend as Backend>::Tx,
        request: &CancelWorkflowRequest,
    ) -> BackendResult<WorkflowExecutionRecord> {
        <RuntimeWorkflowStore as WorkflowStore<InMemoryBackend>>::cancel_tx(
            &self.inner,
            tx,
            request,
        )
    }

    fn inspect(
        &self,
        backend: &InMemoryBackend,
        workflow_execution_id: &str,
    ) -> BackendResult<Option<WorkflowExecutionRecord>> {
        self.inner.inspect(backend, workflow_execution_id)
    }

    fn inspect_tx(
        &self,
        tx: &mut <InMemoryBackend as Backend>::Tx,
        workflow_execution_id: &str,
    ) -> BackendResult<Option<WorkflowExecutionRecord>> {
        <RuntimeWorkflowStore as WorkflowStore<InMemoryBackend>>::inspect_tx(
            &self.inner,
            tx,
            workflow_execution_id,
        )
    }
}

struct LongRunningWorkflowStepHandler;

impl ProvisionServiceV1StepHandler<InMemoryBackend> for LongRunningWorkflowStepHandler {
    fn handle_validate_request(
        &self,
        tx: &mut <InMemoryBackend as Backend>::Tx,
        context: &WorkflowStepHandlerContext,
    ) -> BackendResult<WorkflowStepResult> {
        assert!(
            tx.is_open(),
            "workflow handler received a closed transaction"
        );
        assert_eq!(context.workflow_name, "ProvisionService");
        assert_eq!(context.step_name, "validate_request");
        thread::sleep(Duration::from_millis(850));
        Ok(WorkflowStepResult::complete(None))
    }

    fn handle_create_remote_service(
        &self,
        tx: &mut <InMemoryBackend as Backend>::Tx,
        _context: &WorkflowStepHandlerContext,
    ) -> BackendResult<WorkflowStepResult> {
        assert!(
            tx.is_open(),
            "workflow handler received a closed transaction"
        );
        Ok(WorkflowStepResult::fail(
            "unexpected create_remote_service step",
        ))
    }

    fn handle_wait_for_remote_service(
        &self,
        tx: &mut <InMemoryBackend as Backend>::Tx,
        _context: &WorkflowStepHandlerContext,
    ) -> BackendResult<WorkflowStepResult> {
        assert!(
            tx.is_open(),
            "workflow handler received a closed transaction"
        );
        Ok(WorkflowStepResult::fail(
            "unexpected wait_for_remote_service step",
        ))
    }
}

#[test]
fn long_running_workflow_step_receives_periodic_keepalive() {
    let backend = InMemoryBackend::new();
    let workflows = CountingWorkflowStore::new();

    workflows
        .register_definition(
            &backend,
            &RegisterWorkflowDefinitionRequest {
                definition: WorkflowDefinition {
                    workflow_name: "ProvisionService".to_string(),
                    workflow_version: 1,
                    start_step: "validate_request".to_string(),
                    expected_execution_time: Duration::from_secs(60),
                    singleton: false,
                    steps: vec![
                        WorkflowStepDefinition {
                            name: "validate_request".to_string(),
                            expected_execution_time: Duration::from_secs(1),
                            max_retries: 3,
                        },
                        WorkflowStepDefinition {
                            name: "create_remote_service".to_string(),
                            expected_execution_time: Duration::from_secs(1),
                            max_retries: 3,
                        },
                        WorkflowStepDefinition {
                            name: "wait_for_remote_service".to_string(),
                            expected_execution_time: Duration::from_secs(1),
                            max_retries: 3,
                        },
                    ],
                    metadata: Json::Object(Default::default()),
                },
            },
        )
        .unwrap();
    workflows
        .start(
            &backend,
            &StartWorkflowRequest {
                workflow_execution_id: "wf-long".to_string(),
                workflow_name: "ProvisionService".to_string(),
                workflow_version: 1,
                start_step: "validate_request".to_string(),
                state: Json::Object(Default::default()),
            },
        )
        .unwrap();

    let mut workflow_step_invokers = WorkflowStepInvokerMap::new();
    workflow_provision_service_registry::register_workflow_step_invokers::<InMemoryBackend>(
        &mut workflow_step_invokers,
        std::sync::Arc::new(LongRunningWorkflowStepHandler),
    );
    let runner = WorkflowRunner {
        backend: &backend,
        workflow_store: &workflows,
        workflow_step_invokers: &workflow_step_invokers,
        worker_name: "ProvisionWorker".to_string(),
        lease_duration: Duration::from_secs(1),
        max_attempts: 3,
    };
    let completed = runner
        .run_once("wf-long", "ProvisionService", 1)
        .expect("workflow runner failed")
        .expect("workflow step was not claimed");
    assert_eq!(completed.status, "Completed");
    assert!(
        workflows.keep_alive_count() >= 2,
        "long-running workflow step did not receive periodic keepalive"
    );
}
